src/github.com/getlantern/balancer/balancer.go

```diff
@@ -102,11 +102,17 @@ func (b *Balancer) Dial(network, addr string) (net.Conn, error) {
 		trustedOnly = true
 	}
 
+	var lastDialer *dialer
 	for i := 0; i < dialAttempts; i++ {
 		d, pickErr := b.pickDialer(trustedOnly)
 		if pickErr != nil {
 			return nil, pickErr
 		}
+		if d == lastDialer {
```

Comment:
If one dialer fails or becomes slower than others, the weight will be changed so other dialers take precedence. Yeah, it requires 10s (dial timeout) to notice if a proxy stops response. With the added dial time logging, we can have a better understanding how it behaves from user logs.

Parallelly dialing multiple proxies does work better, but it wastes resourse on both proxy and client, and increases the total file descriptors. A better apprach would be:

1. keep persistent connection to each proxy (disconnect it randomly to be less noticable).
1. keep calculating the throughput and RTT.
1. when new requests come in, forward the request via the fastest proxy, write to a LRU association table.
1. always forward follow-up requests of that domain to same proxy, unless the metrics of that proxy becomes unacceptable.

It requires significant change on both client and proxy, but should be worth doing if the rationale is solid.